### ISSUE TITLE
Follow-up and clean-up for abstract conv2d

### ISSUE THREAD

**Issue by lamblin** [Theano/Theano#3557]
This is a follow-up to the abstract conv2d op mentioned in #2505
Things that remain to be done after #2665:
- [x] Add online documentation about the abstract Op (#3779)
- [ ] Simplify the optimizations introducing GpuCorrMM or CuDNN Ops, by removing the special cases mapping one type of convolution operation to a different one (for instance, mapping gradW to forward) (edit: only when we are guessing the direction, not when it was deliberate for speed, at least for CorrMM)
- [x] Prepare transition from `theano.tensor.nnet.conv.conv2d` to `theano.tensor.nnet.abstract_conv.conv2d` as the default (#3675)
- [x] Port optimizations to the new GPU back-end (~~#3676~~ #3737)
- [x] When #3285 is merged, add optimizations to insert these Ops
- [x] Execute only a subset of tests by default, so the time is not too large (use a numpy RNG seeded with `unittest_tools.fetch_seed()`)
- [x] Update tests to check that the right implementation is inserted by the optimizations (#3668)
- [x] Update (some) tests to call `theano.tensor.nnet.conv2d` (#3732)
- [x] Centralize and unify the implementations of shape inference (#3582)
- [x] Find a way to make test values (and possibly DebugMode) work, but still not enable them in a compiled (optimized) function (#3598)
- [x] Better document "shape" arguments, add runtime checks (#4048)
- [ ] Check what to do about the meta opt?

**Comment by JesseLivezey** [comment]
Would the optimization for the CPU CorrMM look like the ones here?
https://github.com/Theano/Theano/blob/5849db2111a57dbfe6511cc0a12708845aec5daf/theano/sandbox/cuda/opt.py#L2690

I should have time to add them in the next week.

**Comment by lamblin** [comment]
Yes, exactly. You do not have to add the special cases that convert AbstractConv2d to gradI or gradW, in fact I'm arguing for their removal from the existing optimization.
They should simply check for TensorType instead of CudaNdarrayType.
Then, you can register them in a new `LocalGroupDB` along with the existing (legacy) convolution, and register that in `specialize_device` as well. We can give you a hand with that part if needed.
Thanks!

**Comment by fvisin** [comment]
The output's shape in https://github.com/Theano/Theano/blob/3b3b2a4e1d4e932ddd15928d82881524b6f5da01/theano/tensor/nnet/abstract_conv2d.py#L59-L62 should also consider `subsample`

**Comment by fvisin** [comment]
If you want the new interface to work with the legacy code, you should also add an argument `image_shape` (alias for `input_shape`). 

It would also be useful to have an `infer_shape` method (e.g., https://github.com/Theano/Theano/blob/a7cd6648f8a6ceead46919d4abe8047d96909913/theano/tensor/nnet/conv.py#L655). What I suggest is to have a static method that infers the output shape given some parameters and a dynamic method that calls the static one with the attributes of `self`.

**Comment by f0k** [comment]
> Simplify the optimizations introducing GpuCorrMM or CuDNN Ops, by removing the special cases mapping one type of convolution operation to a different one (for instance, mapping gradW to forward)

Note that at least for GpuCorrMM, I'm intentionally using gradW for the forward pass or vice versa if it's faster -- this makes us faster than Caffe for some configurations, it was not just done because we didn't know which one to use. Don't clean up too much :) (And feel free to ask me for reviewing PRs.)

**Comment by lamblin** [comment]
Closing, since I think all that needed to be done was done.
The meta-optimizer is now done inside cuDNN, so the need is not really there any more.

**Comment by f0k** [comment]
> The meta-optimizer is now done inside cuDNN, so the need is not really there any more.

The meta-optimizer is still needed. cuDNN tries different algorithms, but it doesn't always select the optimal code path; sometimes a full convolution forward pass is still faster using the valid convolution backward pass (feel free to [try the examples](http://benanne.github.io/2014/12/09/theano-metaopt.html)). And for the new convolution flavours not supported by cuDNN, there are even more options. But this should be a separate issue, it's not just a follow-up on abstract_conv2d.
